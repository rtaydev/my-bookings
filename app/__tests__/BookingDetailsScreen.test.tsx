import React from "react";
import { render, waitFor, fireEvent, act } from "@testing-library/react-native";
import BookingDetailsScreen from "../booking-details";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import bookingsReducer from "@/store/slices/bookingSlice";

// Mock the expo-router hooks
const mockBack = jest.fn();
jest.mock("expo-router", () => ({
  useLocalSearchParams: jest.fn(() => ({
    bookingId: "ABC123",
    userId: "1",
  })),
  useRouter: jest.fn(() => ({
    back: mockBack,
  })),
}));

// Mock the fetch function to return test booking data
global.fetch = jest.fn(() =>
  Promise.resolve({
    ok: true,
    json: () =>
      Promise.resolve({
        bookingReference: "ABC123",
        title: "Dover to Calais",
        date: "2024-11-05",
        time: "08:00",
        passengers: 2,
        vehicle: "Car",
        details: "Standard vehicle ferry crossing",
        ticketUuid: "8f7e6d5c-4b3a-2a1e-9f8d-7c6b5a4e3d2c",
      }),
  }),
);

const createTestStore = (initialState) => {
  return configureStore({
    reducer: {
      bookings: bookingsReducer,
    },
    preloadedState: {
      bookings: initialState,
    },
  });
};

describe("BookingDetailsScreen", () => {
  const originalError = console.error;
  let store;

  beforeAll(() => {
    console.error = jest.fn();
  });

  beforeEach(() => {
    store = createTestStore({
      booking: null,
      loading: false,
      error: null,
    });

    jest.clearAllMocks();
  });

  afterAll(() => {
    console.error = originalError;
  });

  it("renders QR code correctly", async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <BookingDetailsScreen />
      </Provider>,
    );

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalledWith(
        "http://localhost:3001/users/1/bookings/ABC123",
      );
      expect(getByTestId("qr-code")).toBeTruthy();
    });
  });

  it("handles fetch error", async () => {
    global.fetch.mockImplementationOnce(() =>
      Promise.resolve({
        ok: false,
        status: 404,
        statusText: "Not Found",
      }),
    );

    const { getByTestId } = render(
      <Provider store={store}>
        <BookingDetailsScreen />
      </Provider>,
    );

    await waitFor(() => {
      expect(getByTestId("error-text")).toBeTruthy();
    });
  });

  it("navigates back when back button is pressed", async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <BookingDetailsScreen />
      </Provider>,
    );

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalledWith(
        "http://localhost:3001/users/1/bookings/ABC123",
      );
    });

    await act(async () => {
      const backButton = getByTestId("back-button");
      expect(backButton).toBeTruthy();
      fireEvent.press(backButton);
    });

    expect(mockBack).toHaveBeenCalled();
  });

  it("matches the snapshot", async () => {
    const { toJSON } = render(
      <Provider store={store}>
        <BookingDetailsScreen />
      </Provider>,
    );

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalledWith(
        "http://localhost:3001/users/1/bookings/ABC123",
      );
    });

    expect(toJSON()).toMatchSnapshot();
  });
});
