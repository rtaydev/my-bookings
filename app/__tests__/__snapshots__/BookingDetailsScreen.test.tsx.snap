// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingDetailsScreen matches the snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "flex": 1,
        "padding": 20,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 30,
          "justifyContent": "center",
          "padding": 10,
        },
        {
          "backgroundColor": "transparent",
        },
        {
          "opacity": 1,
        },
        undefined,
        false,
        {
          "alignSelf": "flex-start",
        },
      ]
    }
    testID="back-button"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": undefined,
              "fontSize": 16,
            },
            [
              undefined,
              {
                "color": "#003083",
              },
            ],
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î¨¨
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 16,
            },
            {
              "color": "#003083",
            },
            {
              "marginLeft": 8,
            },
          ]
        }
      >
        Back
      </Text>
    </View>
  </View>
  <RCTScrollView
    style={
      {
        "flex": 1,
        "marginTop": 20,
      }
    }
  >
    <View>
      <Text
        style={
          [
            {
              "fontSize": 24,
              "fontWeight": "bold",
              "marginBottom": 20,
            },
            {
              "color": "#003083",
            },
          ]
        }
      >
        Ferry Reservation Details
      </Text>
      <View
        style={
          [
            {
              "borderRadius": 8,
              "marginBottom": 10,
              "padding": 20,
            },
            {
              "backgroundColor": "#0030831A",
            },
          ]
        }
      >
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Booking Reference:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          ABC123
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Route:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          Dover to Calais
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Date:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          November 5, 2024
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Time:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          08:00
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Passengers:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          2
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Vehicle:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          Car
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "700",
              "marginTop": 8,
            }
          }
        >
          Additional Info:
        </Text>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 18,
            }
          }
        >
          Standard vehicle ferry crossing
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginBottom": 40,
            "marginTop": 40,
          }
        }
      >
        <QRCode
          size={200}
          testID="qr-code"
          value="8f7e6d5c-4b3a-2a1e-9f8d-7c6b5a4e3d2c"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
